### DISCUSSION TITLE
Sometimes facing this exception: `RabbitMQ.Client.Exceptions.ChannelAllocationException`

### DISCUSSION THREAD

**Discussion by ajaykem181** [rabbitmq/rabbitmq-dotnet-client#1637] [Q&A]
<details><summary>Code</summary>

```
using log4net;
using LogzLoggerProvider;
using Microsoft.Extensions.Configuration;
using RabbitMQ.Client;
using System;
using System.Collections.Concurrent;
using System.IO;
using System.Threading;

namespace MarketicEventApi.RabbitMQ
{
    public abstract class RabbitMqClientBase : IDisposable
    {
        private static readonly ILog _logger = LogManager.GetLogger(typeof(ConsumerBase));
        protected string Exchange;
        protected readonly string ExternalSourceQueue;
        protected readonly string ExternalSourceRoutingKey;

        protected readonly string WebsiteQueue;
        protected readonly string WebsiteRoutingKey;
        protected readonly string BookmarkQueue;
        protected readonly string BookmarkRoutingKey;

        protected readonly string TrackEventQueue;
        protected readonly string TrackEventRoutingKey;

        protected readonly string IdentifyEventQueue;
        protected readonly string IdentifyEventRoutingKey;

        protected readonly string ExpiredActionInstanceQueue;
        protected readonly string ExpiredActionInstanceRoutingKey;


        protected IModel Channel { get; private set; }
        private IConnection _connection;
        private readonly ConnectionFactory _connectionFactory;

        protected RabbitMqClientBase(
            ConnectionFactory connectionFactory, IConfiguration configuration)
        {
            Exchange = configuration.GetValue<string>("marketicui:rabbitmq:exchange");
            ExternalSourceQueue = configuration.GetValue<string>("marketicui:rabbitmq:externalSourceQueue");
            ExternalSourceRoutingKey = configuration.GetValue<string>("marketicui:rabbitmq:externalSourceRoutingKey");
            WebsiteQueue = configuration.GetValue<string>("marketicui:rabbitmq:websiteQueue");
            WebsiteRoutingKey = configuration.GetValue<string>("marketicui:rabbitmq:websiteRoutingKey");
            TrackEventQueue = configuration.GetValue<string>("marketicui:rabbitmq:trackEventQueue");
            TrackEventRoutingKey = configuration.GetValue<string>("marketicui:rabbitmq:trackEventRoutingKey");
            IdentifyEventQueue = configuration.GetValue<string>("marketicui:rabbitmq:identifyEventQueue");
            IdentifyEventRoutingKey = configuration.GetValue<string>("marketicui:rabbitmq:identifyEventRoutingKey");
            BookmarkQueue = configuration.GetValue<string>("marketicui:rabbitmq:bookmarkQueue");
            BookmarkRoutingKey = configuration.GetValue<string>("marketicui:rabbitmq:bookmarkRoutingKey");
            ExpiredActionInstanceQueue = configuration.GetValue<string>("marketicui:rabbitmq:expiredactioninstancequeue");
            ExpiredActionInstanceRoutingKey = configuration.GetValue<string>("marketicui:rabbitmq:expiredactioninstanceroutingKey");
            _connectionFactory = connectionFactory;
            ConnectToRabbitMq();
        }

        private void ConnectToRabbitMq()
        {
            _logger.ServiceStarted("Start attempting connection to rabbit mq");

            if (_connection == null || _connection.IsOpen == false)
            {
                _logger.ServiceProcessing("Create connection to rabbit mq");
                _connectionFactory.AutomaticRecoveryEnabled = true;
                _connection = _connectionFactory.CreateConnection();
                _connection.ConnectionShutdown += Connection_ConnectionShutdown;
            }

            _logger.ServiceProcessing("Create channel model and bind queue in rabbit mq");

            if (Channel == null || Channel.IsOpen == false)
            {
                Channel = _connection.CreateModel();
                Channel.ExchangeDeclare(
                    exchange: Exchange,
                    type: "topic",
                    durable: true,
                    autoDelete: false);

                Channel.QueueDeclare(
                    queue: ExternalSourceQueue,
                    durable: true,
                    exclusive: false,
                    autoDelete: false);

                Channel.QueueBind(
                    queue: ExternalSourceQueue,
                    exchange: Exchange,
                    routingKey: ExternalSourceRoutingKey);

                Channel.QueueDeclare(
                    queue: WebsiteQueue,
                    durable: true,
                    exclusive: false,
                    autoDelete: false);

                Channel.QueueBind(
                    queue: WebsiteQueue,
                    exchange: Exchange,
                    routingKey: WebsiteRoutingKey);

                Channel.QueueDeclare(
                    queue: BookmarkQueue,
                    durable: true,
                    exclusive: false,
                    autoDelete: false);

                Channel.QueueBind(
                    queue: BookmarkQueue,
                    exchange: Exchange,
                    routingKey: BookmarkRoutingKey);

                Channel.QueueDeclare(
                    queue: TrackEventQueue,
                    durable: true,
                    exclusive: false,
                    autoDelete: false);

                Channel.QueueBind(
                    queue: TrackEventQueue,
                    exchange: Exchange,
                    routingKey: TrackEventRoutingKey);

                Channel.QueueDeclare(
                    queue: IdentifyEventQueue,
                    durable: true,
                    exclusive: false,
                    autoDelete: false);

                Channel.QueueBind(
                    queue: IdentifyEventQueue,
                    exchange: Exchange,
                    routingKey: IdentifyEventRoutingKey);

                Channel.QueueDeclare(
                    queue: ExpiredActionInstanceQueue,
                    durable: true,
                    exclusive: false,
                    autoDelete: false);

                Channel.QueueBind(
                    queue: ExpiredActionInstanceQueue,
                    exchange: Exchange,
                    routingKey: ExpiredActionInstanceRoutingKey);
            }
        }

        public void Dispose()
        {
            _logger.ServiceProcessing("Rabbit mq connection and channel dispose");

            try
            {
                Channel?.Close();
                Channel?.Dispose();
                Channel = null;

                _connection?.Close();
                _connection?.Dispose();
                _connection = null;
            }
            catch (Exception ex)
            {
                _logger.ServiceFaulted(ex, "Rabbit mq connection and channel dispose failed");
            }
        }

        private void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
        {
            _logger.ServiceProcessing("Rabbit mq connection shutdown event handler called");

            Cleanup();

            while (true)
            {
                try
                {
                    ConnectToRabbitMq();

                    _logger.ServiceFinishedSuccess("Rabbit mq connection reconnected successfully");

                    break;
                }
                catch (Exception ex)
                {
                    _logger.ServiceFaulted(ex, "Rabbit mq reconnection failed");

                    Thread.Sleep(3000);
                }
            }
        }

        private void Cleanup()
        {
           _logger.ServiceProcessing("Close existing rabbit mq channel if exists");

            try
            {
                if (Channel != null && Channel.IsOpen)
                {
                    Channel.Close();
                    Channel = null;
                }

                if (_connection != null && _connection.IsOpen)
                {
                    _connection.Close();
                    _connection = null;
                }
            }
            catch (IOException ex)
            {
                _logger.ServiceFaulted(ex, "Rabbit mq connection close failed");
            }
        }
    }
}
```
</details>

**Comment by lukebakken** [comment]
Your code has duplicated this library's reconnection logic. In `Connection_ConnectionShutdown`, do NOT reconnect.

**Reply by ajaykem181** [reply]
Is that a reason why ChannelAllocationException is coming? or what are the other chances of occurrence of this error?

**Reply by lukebakken** [reply]
Yes, what I said is the most likely explanation. Please fix your code and re-test in your environment.

**Reply by michaelklishin** [reply]
We cannot even guess what is the reason without either reproducing this behavior or at least seeing a stack trace that would show where in library code the exception is thrown.

I can easily see how a `ChannelAllocationException` would be thrown when a connection is first initialized with an allocated channel, then "re-initialized" (the object is replaced) and two different channel objects end up with the same ID (channel number) for a period of time.

In other words, this is an indication of a concurrency hazard affecting the internal state. "Double recovery" is a feasible mistake that would lead to that.

**Reply by ajaykem181** [reply]
1. We are monitoring the application with changes suggested by @lukebakken 
2. We were getting exception when we publish message, please find exception details as below:


RabbitMQ.Client.Exceptions.ChannelAllocationException: The connection cannot support any more channels. Consider creating a new connection
   at RabbitMQ.Client.Impl.SessionManager.Create()
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.CreateNonRecoveringModel()
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.CreateModel()
   at RabbitMQHelper.PooledObjectPolicyImpl.Create()
   at Microsoft.Extensions.ObjectPool.DefaultObjectPool`1.Create()
   at Microsoft.Extensions.ObjectPool.DefaultObjectPool`1.Get()
   at RabbitMQHelper.RabbitMQUtilityImpl.Publish[T](T message, String exchangeName, String exchangeType, String routeKey)

**Reply by ajaykem181** [reply]
I observed one more thing if I forcefully close connection then in next connection multiple channels were created & Idle channels were not closed. Am I missing something in my code? Why Idle channels stays alive for long time?

**Reply by lukebakken** [reply]
Without a way to reproduce this issue, we can't assist you. Please provide code we can compile and run to see this issue.

Team RabbitMQ does _NOT_ have time to guess how to fix the code in your environment.

**Reply by ajaykem181** [reply]
Hi @lukebakken,
I have replicated entire RabbitMQ logic in this project from my original project, once I forcefully close RabbitMQ connection then more than one channels were generated during my testing. Only one channel remained active & running, rest of the channels were Idle specific to this connection for long time & none of the Idle channels were removed automatically.

Also when I used "_connectionFactory.RequestedChannelMax = 1;" in TestWebAPI at that time new Channels were not generated for RabbitMQ connection generated by TestWebAPI. But I couldn't apply same to TestApiMessagePublisher. Am I missing anything in the code because of which Idle channels just stays there?
 
[ConsoleAppRabbitMQ.zip](https://github.com/user-attachments/files/16417999/ConsoleAppRabbitMQ.zip)

**Reply by lukebakken** [reply]
I don't understand what you are trying to demonstrate with this code:

https://github.com/lukebakken/rabbitmq-dotnet-client-1637/blob/main/ConsoleAppRabbitMQ/Program.cs

Of course you are going to run into a channel allocation exception at some point - you are allocating them in a loop.

I deleted the other projects because it was not clear why they were included.

You need to provide better instructions for me to follow.

**Reply by michaelklishin** [reply]
For this exception to be thrown, this function https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/projects/RabbitMQ.Client/util/IntAllocator.cs#L71-L94 must return an error.

@ajaykem181 are you sure you understand what this line does?

```
_connectionFactory.RequestedChannelMax = 1
```

You are telling the library to limit the number of channels that can be opened at any given time to just one. This is pretty unreasonable, the default IIRC is in thousands. If you want to use a lower
number, use 128 or something like that.

**Reply by michaelklishin** [reply]
Yeah, creating channels in a loop without closing them is guaranteed to run into this exception.

@ajaykem181 have you read the [doc guide on channels](https://www.rabbitmq.com/docs/channels#basics)? Including the parts on their expected lifecycle, [resource usage](https://www.rabbitmq.com/docs/channels#resource-usage) and [application-level leaks](https://www.rabbitmq.com/docs/channels#monitoring)?

**Reply by ajaykem181** [reply]
@lukebakken , @michaelklishin ,
 Sorry, I didn't mention that TestWebAPI is acknowledging messages & TestApiMessagePublisher has a backgroundworker which publish messages. 
I had removed ConsoleAppRabbitMQ from solution, it was just there to check if ChannelAllocationException exception occurs when channels limit 2047 is crossed & then i started searching for solution, as same error was logged in my application.

**Reply by ajaykem181** [reply]
> I don't understand what you are trying to demonstrate with this code:
> 
> https://github.com/lukebakken/rabbitmq-dotnet-client-1637/blob/main/ConsoleAppRabbitMQ/Program.cs
> 
> Of course you are going to run into a channel allocation exception at some point - you are allocating them in a loop.
> 
> I deleted the other projects because it was not clear why they were included.
> 
> You need to provide better instructions for me to follow.

Please do not follow this Console App, consider other two API applications as I mentioned in earlier comment. 
[ConsoleAppRabbitMQ.zip](https://github.com/user-attachments/files/16422419/ConsoleAppRabbitMQ.zip)

**Reply by lukebakken** [reply]
> consider other two API applications as I mentioned in earlier comment

How do I run and use them, and then test your scenario? I do _NOT_ have the time to guess.

I will only investigate if you give me precise instructions.

**Reply by ajaykem181** [reply]
Its created on dot net 6.0, If you are able to build the solution. Then both applications should be executed together, BackgroundService in  TestApiMessagePublisher will publish messages TestWebAPI will consume messages.

**Reply by lukebakken** [reply]
And then what? Forcefully close a connection? Please take the time to be precise and list EVERY step.

Also, please see this comment - https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1637#discussioncomment-10192193

I can't build your code.

**Comment by lukebakken** [comment]
@ajaykem181 your code from [this comment](https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1637#discussioncomment-10186159) does not build:

```
C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637 [main ≡]
> dotnet build .\ConsoleAppRabbitMQ.sln
  Determining projects to restore...
C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637\TestApiMessagePublisher\TestApiMessagePublisher.csproj : error NU1101: Unable to find package RabbitMQHelper. No packages exist with this id in source(s): Microsoft Visu
al Studio Offline Packages, nuget.org [C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637\ConsoleAppRabbitMQ.sln]
C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637\TestWebAPI\TestWebAPI.csproj : error NU1101: Unable to find package RabbitMQHelper. No packages exist with this id in source(s): Microsoft Visual Studio Offline Packages
, nuget.org [C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637\ConsoleAppRabbitMQ.sln]
  Failed to restore C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637\TestApiMessagePublisher\TestApiMessagePublisher.csproj (in 491 ms).
  Failed to restore C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637\TestWebAPI\TestWebAPI.csproj (in 491 ms).

Build FAILED.

C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637\TestApiMessagePublisher\TestApiMessagePublisher.csproj : error NU1101: Unable to find package RabbitMQHelper. No packages exist with this id in source(s): Microsoft Visu
al Studio Offline Packages, nuget.org [C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637\ConsoleAppRabbitMQ.sln]
C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637\TestWebAPI\TestWebAPI.csproj : error NU1101: Unable to find package RabbitMQHelper. No packages exist with this id in source(s): Microsoft Visual Studio Offline Packages
, nuget.org [C:\Users\lbakken\development\lukebakken\rabbitmq-dotnet-client-1637\ConsoleAppRabbitMQ.sln]
    0 Warning(s)
    2 Error(s)
```

I have checked in your latest code to this repository:

https://github.com/lukebakken/rabbitmq-dotnet-client-1637

Please fork my repository, make the necessary fixes, and open a pull request so that I can merge them.
